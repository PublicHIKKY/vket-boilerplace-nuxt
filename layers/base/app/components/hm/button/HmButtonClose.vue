<template>
  <HaBaseButton
    class="hm-button-close"
    @click="onClick"
  >
    <IconClose
      :style="{
        width: width,
        height: height,
      }"
    />
  </HaBaseButton>
</template>

<script setup lang="ts">
import IconClose from '#base/app/assets/icons/icon-close.svg'

type Emits = {
  (e: 'click'): () => void
}

const emit = defineEmits<Emits>()
const onClick = () => emit('click')

type Props = {
  width?: string
  height?: string
}

withDefaults(defineProps<Props>(), {
  width: '20px',
  height: '20px',
})
</script>

<style lang="scss" scoped>
@use '#base/app/assets/styles/variables' as v;
@use 'sass:color';

.hm-button-close {
  :deep(circle) {
    transition: fill 0.2s ease;
  }

  &:hover :deep(circle) {
    // note: opacityだと後ろのborderが見えてしまうのでsvgの色を変更
    fill: color.adjust(v.$gray-1, $lightness: 5%);
  }
}
</style>
