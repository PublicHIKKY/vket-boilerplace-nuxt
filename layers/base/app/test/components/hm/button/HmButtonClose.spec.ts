import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import HmButtonClose from '#base/app/components/hm/button/HmButtonClose.vue'

test('ref component', () => {
  expect(HmButtonClose).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HmButtonClose)
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

describe('props', () => {
  describe(':width', () => {
    test('default is 20px', () => {
      const wrapper = mount(HmButtonClose)
      expect(wrapper.get('svg').attributes().style).toContain(`width: 20px`)
    })
    // 適当な値を渡して、その数値と同じwidthになっているかテストする
    test('pass prop', () => {
      const width = Math.floor(Math.random() * 101)
      const wrapper = mount(HmButtonClose, { props: { width: `${width}px` } })
      expect(wrapper.get('svg').attributes().style).toContain(
        `width: ${width}px`,
      )
    })
  })
  describe(':height', () => {
    test('default is 20px', () => {
      const wrapper = mount(HmButtonClose)
      expect(wrapper.get('svg').attributes().style).toContain(`height: 20px`)
    })
    test('pass prop', () => {
      const height = Math.floor(Math.random() * 101)
      const wrapper = mount(HmButtonClose, { props: { height: `${height}px` } })
      expect(wrapper.get('svg').attributes().style).toContain(
        `height: ${height}px`,
      )
    })
  })
})

describe('emits', () => {
  test('click emits clickEvent', async () => {
    const wrapper = mount(HmButtonClose)
    await wrapper.get('button').trigger('click')
    expect(wrapper.emitted().click?.length).toBe(1)
  })
})
