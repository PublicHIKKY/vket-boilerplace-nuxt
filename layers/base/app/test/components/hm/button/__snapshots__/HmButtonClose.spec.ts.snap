// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `
"<button data-v-0844060a="" data-v-50f49dd3="" class="ha-base-button hm-button-close" type="button"><svg data-v-50f49dd3="" xmlns="http://www.w3.org/2000/svg" width="24.121" height="24.121" viewBox="0 0 24.121 24.121" style="width: 20px; height: 20px;">
    <g transform="translate(-953.439 -212.439)">
      <line x2="22" y2="22" transform="translate(954.5 213.5)" fill="none" stroke="#fff" stroke-width="3"></line>
      <line x1="22" y2="22" transform="translate(954.5 213.5)" fill="none" stroke="#fff" stroke-width="3"></line>
    </g>
  </svg></button>"
`;
